### PR TITLE
core: implement core_mutex_priority_inheritance

### PR DESCRIPTION
**Pull request by maribu**
### Contribution description

This implements `core_mutex_mitigate_priority_inversion`, a module that employs the textbook solution of priority inheritance to mitigate priority inversion.

In addition, the corresponding test was cleaned up and automated:
- drop timer dependency, just use mutex to stage the priority inversion scenario
- reduce memory consumption so that `Makefile.ci` can be dropped by using `fmt` over stdio and reducing stack sizes
- reduce the complexity of the scenario to the bare minimum required to produce the issue

### Testing procedure

```
make flash test -C tests/thread_priority_inversion
```

should build, flash and run. It should fail without `core_mutex_mitigate_priority_inversion` (comment out line in the test's `Makefile`), but pass with if it is used.

### Issues/PRs references

None

### CODE REVIEW

**Comment by OlegHahm** [comment]
Where are we at here?

**Comment by maribu** [comment]
The infrastructure to change the priority of a running thread is now in place. So half way there. (Or actually even more, extending the mutex to make use of it is relatively simple.) I think I can PR this today.

**Comment by maribu** [comment]
I added the corresponding code in a new commit to this PR. It kind of makes sense to me have the fix in the corresponding test application (or rather it's cleanup) in the same PR.

**Comment by benpicco** [comment]
Just wondering, why the clunky name? Isn't this just `core_mutex_priority_inheritance`?

**Comment by maribu** [comment]
My reasoning is that priority inheritance is just the implementation to mitigate priority inversion. There are [other possible solutions](https://en.wikipedia.org/wiki/Priority_inversion#Solutions) as well.

I think in the end the user will care little how the priority inversion is mitigated, as long as the result is that priority inversions no longer occur. And maybe a future PR changes the solution to something that has a better trade-off. With the current name, we could just keep the module name.

**Comment by benpicco** [comment]
Sure but this particular implementation is priority inheritance. If someone knows about *priority inversion*, they will most likely also know about *priority inheritance*.

To me such a long module name looks like something a user should not touch / that enables an ugly workaround and thus needs a lot of explaining. But priority inheritance is already a well established term, no need to describe it as "priority inversion mitigation" - a user might actually want to know what they are getting there.

But enough bikeshedding, if you like the name, stick with it, I just think it's a bit ugly.

File: core/mutex.c [hunk 1]
```diff
@@ -136,6 +154,16 @@ void mutex_unlock(mutex_t *mutex)
...
 
+#ifdef MODULE_CORE_MUTEX_MITIGATE_PRIORITY_INVERSION
+    thread_t *owner = mutex->owner;
+    if ((owner) && (owner->priority != mutex->owner_original_priority)) {
+        DEBUG("PID[%" PRIkernel_pid "] prio %u --> %u\n",
+              owner->pid,
+              (unsigned)owner->priority, (unsigned)mutex->owner->priority);
+        sched_change_priority(owner, mutex->owner_original_priority);
+    }
+#endif
```
Comment (benpicco) [inline core/mutex.c:165]: Should we set `mutex->owner = NULL;` here?

**Review by benpicco — APPROVED** [review]
Now that RIOT is going to space, this is surely an important feature to have :wink:

**Comment by benpicco** [comment]
> what's the difference to #7445?

I makes use of an API that did not exist 4 years ago and is not stale?

**Comment by maribu** [comment]
Note that decoupling priority inheritance for mutex and msg makes sense, as there is a trade-off. In a use case that needs to mitigate priority inversion for shared bus accesses is not automatically affected by priority inversion via msg.

The issue of priority inversion can occur due to hardware limitations (e.g. a low prio and a high prio task needing to talk to hardware on the single I2C bus available) and for some use cases cannot be avoided. I have to admit that I didn't spent much time thinking about it, but since msg is not linked to shared hardware resource, I believe that priority inversion could be solved by carefully designing the software so that it cannot occur.

In any case, https://github.com/RIOT-OS/RIOT/pull/7445 implements priority inheritance only for `msg_send_receive()`, not for `msg_send()` or `msg_receive()`. And even there the caller of `msg_send_receive()` would have to first to potentially boost the priority of the receiver to not block on sending for longer than expected, and afterwards boost the priority of the sender to not block longer than expected on receiving. So 25 % of the places where priority inheritance would be needed are actually implemented in https://github.com/RIOT-OS/RIOT/pull/7445 - and given that it is stale for quite some time, I don't expect this to be fixed swiftly.

We can still add priority inheritance for msg later on, if that is indeed needed.

**Comment by benpicco** [comment]
C++ does not like the mutex initialization macro 😕

**Comment by maribu** [comment]
Ah, yes. Named initializers (`{ .name = 42 }`) area C++2x feature :( Let me see if I can find an initializer that both C and C++ like.

**Comment by maribu** [comment]
The use of `fwrite()` to fix output when modules are intermixing `fmt` and `stdio` lead to stack overflows in the test here now :/

I wonder if we should just provide a tiny stdio implementation and make that a default module when `newlib` is in use :/

**Comment by maribu** [comment]
I changed the `mutex_t` to store the pid rather than a pointer to the thread to safe a bit of RAM. Locally, the `rmutex_t` now fits in the ESP32's `sturct __lock`. Let's see if Murdock also likes it.

At least this forced me to re-run the test and notice that I had to increase the stack sizes due to https://github.com/RIOT-OS/RIOT/pull/18162 being merged.

**Comment by benpicco** [comment]
Please squash!

**Comment by maribu** [comment]
All green and test is still passing

**Comment by maribu** [comment]
Uncrustify should now also hopefully be happy

**Review by kfessel — APPROVED** [review]
I trust the test run by @maribu and @benpicco
this is a feature long asked for (as indicated by a similar PR created in 2017) - lets check that box

**Comment by maribu** [comment]
thx :)

**Comment by MrKevinWeiss** [comment]
Hmm just adding the `core_mutex_priority_inheritance` seems to break the `esp32-wroom-32` board.  I am trying to investigate.

**Pull request merged**